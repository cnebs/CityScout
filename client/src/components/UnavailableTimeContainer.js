import React from 'react';
import { List } from '@material-ui/core';
import UnavailableTimeIndividual from './UnavailableTimeIndividual';

export default function UnavailableTimeContainer({ times, handleDelete }) {
  console.log(times);
  return (
    <List>
      {times.map(time => (
        <UnavailableTimeIndividual time={time} handleDelete={handleDelete} />
      ))}
    </List>
  );
}
