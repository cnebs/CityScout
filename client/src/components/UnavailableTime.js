import React from 'react';
import {
  Grid,
  Typography,
  Fab,
  TextField,
  FormControl,
  InputLabel,
  Select,
  MenuItem
} from '@material-ui/core/';
import AddIcon from '@material-ui/icons/Add';
import DateFnsUtils from '@date-io/date-fns';
import {
  MuiPickersUtilsProvider,
  KeyboardTimePicker
} from '@material-ui/pickers';
import moment from 'moment';
import axios from 'axios';
import UnavailableTimeContainer from './UnavailableTimeContainer';
import Title from './Title.js';

export default class UnavailableTime extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      date: moment().format('YYYY-MM-DD'),
      timeStart: moment().format('HH:mm:00'),
      timeEnd: moment().format('HH:mm:00'),
      eventName: '',
      daySelected: '',
      unavailableTimes: []
    };
    this.handleTimeStartChange = this.handleTimeStartChange.bind(this);
    this.handleTimeEndChange = this.handleTimeEndChange.bind(this);
    this.handleDayOfWeekChange = this.handleDayOfWeekChange.bind(this);
    this.handleNameChange = this.handleNameChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.getUnavailableTimes = this.getUnavailableTimes.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }

  handleTimeStartChange(_, value) {
    const timeStart = moment(value, 'HH:mm a').format('HH:mm:00');
    this.setState({ timeStart });
  }

  handleTimeEndChange(_, value) {
    const timeEnd = moment(value, 'HH:mm a').format('HH:mm:00');
    this.setState({ timeEnd });
  }

  handleDayOfWeekChange(e) {
    const daySelected = e.target.value;
    const date = moment()
      .isoWeekday(daySelected)
      .format('YYYY-MM-DD');
    this.setState({ date, daySelected });
  }

  handleSubmit() {
    axios
      .post(`/api/unavailable`, {
        token: this.props.userToken,
        name: this.state.eventName,
        time_start: moment(
          `${this.state.date} ${this.state.timeStart}`
        ).toISOString(),
        time_end: moment(
          `${this.state.date} ${this.state.timeEnd}`
        ).toISOString()
      })
      .then(_ => this.getUnavailableTimes())
      .catch();
  }

  handleDelete(item_id) {
    axios
      .delete(
        `/api/unavailable?token=${this.props.userToken}&item_id=${item_id}`
      )
      .then(_ => {
        this.setState(state => {
          return {
            unavailableTimes: state.unavailableTimes.filter(
              time => time.item_id !== item_id
            )
          };
        });
      })
      .catch();
  }

  handleNameChange(eventName) {
    this.setState({ eventName });
  }

  getUnavailableTimes() {
    axios
      .get(`/api/unavailable?token=${this.props.userToken}`)
      .then(({ data }) => data.filter(time => time.recurring))
      .then(unavailableTimes => this.setState({ unavailableTimes }))
      .catch();
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.userToken !== prevProps.userToken) {
      this.getUnavailableTimes();
    }
  }

  componentDidMount() {
    this.getUnavailableTimes();
  }

  render() {
    return (
      <>
        <Title>When are you busy?</Title>
        <Typography variant="subtitle2" color="textSecondary">
          Tell us when you are unavailable. We assume these times are recurring.
          Please delete them below if they are no longer relevant!
        </Typography>

        <Grid container>
          <Grid item xs={8}>
            <TextField
              id="standard-name"
              label="Event Name"
              defaultValue=""
              onChange={e => this.handleNameChange(e.target.value)}
              style={{ minWidth: 250 }}
            />
          </Grid>
          <Grid item xs={4}>
            <form autoComplete="off">
              <FormControl style={{ minWidth: 120 }}>
                <InputLabel htmlFor="dayOfTheWeek">Days</InputLabel>
                <Select
                  onChange={e => this.handleDayOfWeekChange(e)}
                  inputProps={{
                    name: 'day',
                    id: 'dayOfTheWeek'
                  }}
                  value={this.state.daySelected}
                >
                  <MenuItem value={0}>Sundays</MenuItem>
                  <MenuItem value={1}>Mondays</MenuItem>
                  <MenuItem value={2}>Tuesdays</MenuItem>
                  <MenuItem value={3}>Wednesdays</MenuItem>
                  <MenuItem value={4}>Thursdays</MenuItem>
                  <MenuItem value={5}>Fridays</MenuItem>
                  <MenuItem value={6}>Saturdays</MenuItem>
                </Select>
              </FormControl>
            </form>
          </Grid>

          <MuiPickersUtilsProvider utils={DateFnsUtils}>
            <Grid item xs={12} sm={6}>
              <KeyboardTimePicker
                margin="normal"
                id="time-picker"
                label="Time Start"
                value={this.state.datepickerStart}
                onChange={this.handleTimeStartChange}
                minutesStep={5}
                KeyboardButtonProps={{
                  'aria-label': 'change time'
                }}
              />
            </Grid>
            <Grid item xs={12} sm={6}>
              <KeyboardTimePicker
                margin="normal"
                id="time-picker"
                label="Time End"
                value={this.state.datepickerEnd}
                onChange={this.handleTimeEndChange}
                minutesStep={5}
                KeyboardButtonProps={{
                  'aria-label': 'change time'
                }}
              />
            </Grid>
          </MuiPickersUtilsProvider>
        </Grid>

        <Fab color="primary" aria-label="add">
          <AddIcon onClick={() => this.handleSubmit()} />
        </Fab>
        <UnavailableTimeContainer
          times={this.state.unavailableTimes}
          handleDelete={this.handleDelete}
        />
      </>
    );
  }
}
