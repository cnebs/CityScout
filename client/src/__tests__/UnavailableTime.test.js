import { shallow } from 'enzyme';
import React from 'react';
import UnavailableTime from '../components/UnavailableTime';
import { Grid, Typography, Fab, TextField } from '@material-ui/core/';
import { KeyboardTimePicker, KeyboardDatePicker } from '@material-ui/pickers';
import UnavailableTimeContainer from '../components/UnavailableTimeContainer';

describe('UnavailableTime component', () => {
  let wrapper;
  // const getUnavailableTimes = jest.fn();
  beforeAll(() => {
    wrapper = shallow(<UnavailableTime userToken="foo" />, {
      lifecycleExperimental: true
    });
  });
  test('it renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('it renders the corect child components', () => {
    expect(wrapper.find(UnavailableTimeContainer).length).toBe(1);
    expect(wrapper.find(Grid).length).toBeGreaterThan(0);
    expect(wrapper.find(Typography).length).toBeGreaterThan(0);
    expect(wrapper.find(TextField).length).toBe(1);
    expect(wrapper.find(Fab).length).toBe(1);
    expect(wrapper.find(KeyboardDatePicker).length).toBe(1); // only 1 DatePicker
    expect(wrapper.find(KeyboardTimePicker).length).toBe(2); // 2 TimePickers should render
  });
  // test('it calls getUnavailableTimes when props change', () => {
  //   wrapper.getUnavailableTimes = jest.fn();
  //   wrapper.setProps({ userToken: 'bar' });
  //   expect(wrapper.instance().getUnavailableTimes).toBeCalled();
  // });
});
