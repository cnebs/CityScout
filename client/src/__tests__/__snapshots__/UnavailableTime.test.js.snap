// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnavailableTime component it renders without crashing 1`] = `ShallowWrapper {}`;
